File: SaidIt/src/main/java/eu/mrogalski/saidit/SaidItFragment.java
package eu.mrogalski.saidit;

import android.app.Activity;
import android.app.Notification;
import android.app.PendingIntent;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.net.Uri;
import android.os.Bundle;
import android.os.IBinder;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.core.app.ActivityCompat;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.core.content.FileProvider;
import androidx.fragment.app.Fragment;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.button.MaterialButtonToggleGroup;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textview.MaterialTextView;

import java.io.File;

import eu.mrogalski.android.TimeFormat;

public class SaidItFragment extends Fragment implements SaveClipBottomSheet.SaveClipListener {

    private static final String YOUR_NOTIFICATION_CHANNEL_ID = "SaidItServiceChannel";
    private SaidItService echo;

    // UI Elements
    private View recordingGroup;
    private View listeningGroup;
    private MaterialTextView recordingTime;
    private MaterialTextView historySize;
    private MaterialButtonToggleGroup listeningToggleGroup;

    // State
    private boolean isRecording = false;
    private float memorizedDuration = 0;

    private final Runnable updater = new Runnable() {
        @Override
        public void run() {
            if (getView() == null || echo == null) return;
            echo.getState(serviceStateCallback);
        }
    };

    private final ServiceConnection echoConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName className, IBinder binder) {
            SaidItService.BackgroundRecorderBinder typedBinder = (SaidItService.BackgroundRecorderBinder) binder;
            echo = typedBinder.getService();
            if (getView() != null) {
                getView().postOnAnimation(updater);
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName arg0) {
            echo = null;
        }
    };

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_background_recorder, container, false);
        final Activity activity = requireActivity();

        // Find new UI elements
        recordingGroup = rootView.findViewById(R.id.recording_group);
        listeningGroup = rootView.findViewById(R.id.listening_group);
        recordingTime = rootView.findViewById(R.id.recording_time);
        historySize = rootView.findViewById(R.id.history_size);
        MaterialButton saveClipButton = rootView.findViewById(R.id.save_clip_button);
        MaterialButton settingsButton = rootView.findViewById(R.id.settings_button);
        MaterialButton stopRecordingButton = rootView.findViewById(R.id.rec_stop_button);
        listeningToggleGroup = rootView.findViewById(R.id.listening_toggle_group);

        // Set listeners
        settingsButton.setOnClickListener(v -> startActivity(new Intent(activity, SettingsActivity.class)));

        stopRecordingButton.setOnClickListener(v -> {
            if (echo != null) {
                echo.stopRecording(new PromptFileReceiver(activity), "");
            }
        });

        saveClipButton.setOnClickListener(v -> {
            SaveClipBottomSheet bottomSheet = SaveClipBottomSheet.newInstance(memorizedDuration);
            bottomSheet.setSaveClipListener(this);
            bottomSheet.show(getParentFragmentManager(), "SaveClipBottomSheet");
        });

        listeningToggleGroup.addOnButtonCheckedListener((group, checkedId, isChecked) -> {
            if (isChecked && echo != null) {
                if (checkedId == R.id.listening_button) {
                    echo.enableListening();
                } else if (checkedId == R.id.disabled_button) {
                    echo.disableListening();
                }
            }
        });

        return rootView;
    }

    @Override
    public void onSaveClip(String fileName, float durationInSeconds) {
        if (echo != null) {
            AlertDialog progressDialog = new MaterialAlertDialogBuilder(requireActivity())
                    .setTitle("Saving Recording")
                    .setMessage("Please wait...")
                    .setCancelable(false)
                    .create();
            progressDialog.show();

            echo.dumpRecording(durationInSeconds, new PromptFileReceiver(getActivity(), progressDialog), fileName);
        }
    }


    private final SaidItService.StateCallback serviceStateCallback = new SaidItService.StateCallback() {
        @Override
        public void state(final boolean listeningEnabled, final boolean recording, final float memorized, final float totalMemory, final float recorded) {
            memorizedDuration = memorized;
            
            if (isRecording != recording) {
                isRecording = recording;
                recordingGroup.setVisibility(recording ? View.VISIBLE : View.GONE);
                listeningGroup.setVisibility(recording ? View.GONE : View.VISIBLE);
            }

            if (isRecording) {
                recordingTime.setText(TimeFormat.shortTimer(recorded));
            } else {
                historySize.setText(TimeFormat.shortTimer(memorized));
            }

            // Update listening toggle state without triggering listener
            listeningToggleGroup.removeOnButtonCheckedListener(null);
            if (listeningEnabled) {
                listeningToggleGroup.check(R.id.listening_button);
                listeningGroup.setAlpha(1.0f);
            } else {
                listeningToggleGroup.check(R.id.disabled_button);
                listeningGroup.setAlpha(0.5f);
            }

            if (getView() != null) {
                getView().postOnAnimationDelayed(updater, 100);
            }
        }
    };

    @Override
    public void onStart() {
        super.onStart();
        final Activity activity = requireActivity();
        activity.bindService(new Intent(activity, SaidItService.class), echoConnection, Context.BIND_AUTO_CREATE);
    }

    @Override
    public void onStop() {
        super.onStop();
        requireActivity().unbindService(echoConnection);
        echo = null;
    }

    // --- File Receiver and Notification Logic ---

    static Notification buildNotificationForFile(Context context, File outFile) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        Uri fileUri = FileProvider.getUriForFile(context, context.getApplicationContext().getPackageName() + ".provider", outFile);
        intent.setDataAndType(fileUri, "audio/wav");
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);

        PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_IMMUTABLE);

        NotificationCompat.Builder notificationBuilder = new NotificationCompat.Builder(context, YOUR_NOTIFICATION_CHANNEL_ID)
                .setContentTitle(context.getString(R.string.recording_saved))
                .setContentText(outFile.getName())
                .setSmallIcon(R.drawable.ic_stat_notify_recorded)
                .setTicker(outFile.getName())
                .setContentIntent(pendingIntent)
                .setAutoCancel(true);
        notificationBuilder.setPriority(NotificationCompat.PRIORITY_DEFAULT);
        return notificationBuilder.build();
    }

    static class NotifyFileReceiver implements SaidItService.WavFileReceiver {
        private final Context context;

        public NotifyFileReceiver(Context context) {
            this.context = context;
        }

        @Override
        public void fileReady(final File file, float runtime) {
            NotificationManagerCompat notificationManager = NotificationManagerCompat.from(context);
            if (ActivityCompat.checkSelfPermission(context, android.Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
                return;
            }
            notificationManager.notify(43, buildNotificationForFile(context, file));
        }
    }


    static class PromptFileReceiver implements SaidItService.WavFileReceiver {
        private final Activity activity;
        private final AlertDialog progressDialog;

        public PromptFileReceiver(Activity activity, AlertDialog dialog) {
            this.activity = activity;
            this.progressDialog = dialog;
        }

        public PromptFileReceiver(Activity activity) {
            this(activity, null);
        }

        @Override
        public void fileReady(final File file, float runtime) {
            if (progressDialog != null && progressDialog.isShowing()) {
                progressDialog.dismiss();
            }
            if (activity != null && !activity.isFinishing()) {
                Uri fileUri = FileProvider.getUriForFile(activity, activity.getApplicationContext().getPackageName() + ".provider", file);
                new MaterialAlertDialogBuilder(activity)
                        .setTitle(R.string.recording_done_title)
                        .setMessage(file.getName())
                        .setPositiveButton(R.string.open, (dialog, which) -> {
                            Intent intent = new Intent(Intent.ACTION_VIEW);
                            intent.setDataAndType(fileUri, "audio/wav");
                            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                            activity.startActivity(intent);
                        })
                        .setNeutralButton(R.string.share, (dialog, which) -> {
                            Intent shareIntent = new Intent(Intent.ACTION_SEND);
                            shareIntent.setType("audio/wav");
                            shareIntent.putExtra(Intent.EXTRA_STREAM, fileUri);
                            shareIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                            activity.startActivity(Intent.createChooser(shareIntent, "Send to"));
                        })
                        .setNegativeButton(R.string.dismiss, null)
                        .show();
            }
        }
    }
}

File: SaidIt/src/main/kotlin/eu/mrogalski/saidit/SaidItFragment.kt
package eu.mrogalski.saidit

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.VisibleForTesting
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.getkeepsafe.taptargetview.TapTarget
import com.getkeepsafe.taptargetview.TapTargetSequence
import com.google.android.material.button.MaterialButton
import com.google.android.material.button.MaterialButtonToggleGroup
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.textview.MaterialTextView
import com.siya.epistemophile.R
import com.siya.epistemophile.features.recorder.RecordingUiState
import com.siya.epistemophile.features.recorder.RecordingViewModel
import dagger.hilt.android.AndroidEntryPoint
import eu.mrogalski.android.TimeFormat
import javax.inject.Inject
import kotlinx.coroutines.launch

@AndroidEntryPoint
class SaidItFragment : Fragment(), SaveClipBottomSheet.SaveClipListener {

    companion object {
        private const val TOUR_START_DELAY = 500L
    }

    @Inject
    lateinit var viewModel: RecordingViewModel

    // UI Elements
    private var recordingGroup: View? = null
    private var listeningGroup: View? = null
    private var recordingTime: MaterialTextView? = null
    private var historySize: MaterialTextView? = null
    private var listeningToggleGroup: MaterialButtonToggleGroup? = null

    // State
    private var isRecording = false
    private var memorizedDuration = 0f

    @VisibleForTesting
    internal var progressDialogForTest: AlertDialog? = null

    @VisibleForTesting
    internal val listeningToggleListener = MaterialButtonToggleGroup.OnButtonCheckedListener { _, checkedId, isChecked ->
        if (isChecked) {
            when (checkedId) {
                R.id.listening_button -> viewModel.toggleListening(true)
                R.id.disabled_button -> viewModel.toggleListening(false)
            }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val rootView = inflater.inflate(R.layout.fragment_background_recorder, container, false)
        val activity = requireActivity()

        // Find UI elements
        val toolbar: Toolbar = rootView.findViewById(R.id.toolbar)
        recordingGroup = rootView.findViewById(R.id.recording_group)
        listeningGroup = rootView.findViewById(R.id.listening_group)
        recordingTime = rootView.findViewById(R.id.recording_time)
        historySize = rootView.findViewById(R.id.history_size)
        val saveClipButton: MaterialButton = rootView.findViewById(R.id.save_clip_button)
        val settingsButton: MaterialButton = rootView.findViewById(R.id.settings_button)
        val recordingsButton: MaterialButton = rootView.findViewById(R.id.recordings_button)
        val stopRecordingButton: MaterialButton = rootView.findViewById(R.id.rec_stop_button)
        listeningToggleGroup = rootView.findViewById(R.id.listening_toggle_group)

        toolbar.setOnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.action_help -> {
                    startActivity(android.content.Intent(requireActivity(), HowToActivity::class.java))
                    true
                }
                else -> false
            }
        }

        settingsButton.setOnClickListener {
            startActivity(android.content.Intent(activity, SettingsActivity::class.java))
        }

        recordingsButton.setOnClickListener {
            startActivity(android.content.Intent(activity, RecordingsActivity::class.java))
        }

        stopRecordingButton.setOnClickListener { viewModel.stopRecording() }

        saveClipButton.setOnClickListener {
            val bottomSheet = SaveClipBottomSheet.newInstance(memorizedDuration)
            bottomSheet.setSaveClipListener(this)
            bottomSheet.show(parentFragmentManager, "SaveClipBottomSheet")
        }

        listeningToggleGroup?.addOnButtonCheckedListener(listeningToggleListener)

        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch { viewModel.uiState.collect { renderUi(it) } }
            }
        }
    }

    private fun renderUi(state: RecordingUiState) {
        val recording = state.isRecording
        if (isRecording != recording) {
            isRecording = recording
            recordingGroup?.visibility = if (recording) View.VISIBLE else View.GONE
            listeningGroup?.visibility = if (recording) View.GONE else View.VISIBLE
        }

        // Update timers from ViewModel state
        memorizedDuration = state.memorizedSeconds
        if (recording) {
            recordingTime?.text = TimeFormat.shortTimer(state.recordedSeconds)
        } else {
            historySize?.text = TimeFormat.shortTimer(state.memorizedSeconds)
        }

        // Update listening toggle without bouncing the listener
        listeningToggleGroup?.removeOnButtonCheckedListener(listeningToggleListener)
        if (state.isListening) {
            listeningToggleGroup?.check(R.id.listening_button)
            listeningGroup?.alpha = 1.0f
        } else {
            listeningToggleGroup?.check(R.id.disabled_button)
            listeningGroup?.alpha = 0.5f
        }
        listeningToggleGroup?.addOnButtonCheckedListener(listeningToggleListener)
    }

    override fun onSaveClip(fileName: String, durationInSeconds: Float) {
        val progressDialog = MaterialAlertDialogBuilder(requireActivity())
            .setTitle("Saving Recording")
            .setMessage("Please wait...")
            .setCancelable(false)
            .create()
        progressDialog.setOnDismissListener { progressDialogForTest = null }
        progressDialog.show()
        progressDialogForTest = progressDialog

        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.saveRecording(durationInSeconds, fileName)
            if (progressDialog.isShowing) progressDialog.dismiss()
        }
    }

    @VisibleForTesting
    fun isProgressDialogShowingForTest(): Boolean = progressDialogForTest?.isShowing == true

    @VisibleForTesting
    fun getProgressDialogForTest(): AlertDialog? = progressDialogForTest

    fun startTour() {
        // A small delay to ensure the UI is fully drawn before starting the tour.
        view?.postDelayed(::startInteractiveTour, TOUR_START_DELAY)
    }

    private fun startInteractiveTour() {
        val currentActivity = activity ?: return
        val currentView = view ?: return

        val sequence = TapTargetSequence(currentActivity)
        sequence.targets(
            TapTarget.forView(currentView.findViewById<View>(R.id.listening_toggle_group),
                getString(R.string.tour_listening_toggle_title),
                getString(R.string.tour_listening_toggle_desc))
                .cancelable(false).tintTarget(false),
            TapTarget.forView(currentView.findViewById<View>(R.id.history_size),
                getString(R.string.tour_memory_holds_title),
                getString(R.string.tour_memory_holds_desc))
                .cancelable(false).tintTarget(false),
            TapTarget.forView(currentView.findViewById<View>(R.id.save_clip_button),
                getString(R.string.tour_save_clip_title),
                getString(R.string.tour_save_clip_desc))
                .cancelable(false).tintTarget(false),
            TapTarget.forView(currentView.findViewById<View>(R.id.bottom_buttons_layout),
                getString(R.string.tour_bottom_buttons_title),
                getString(R.string.tour_bottom_buttons_desc))
                .cancelable(false).tintTarget(false)
        )
        sequence.start()
    }
}

